### PR TITLE
Slasher protection race condition fix 

### LINKED ISSUE #6868: Slasher race condition can lead to slashing
# 🐞 Bug Report

### Description

When sending multiple requests to the slasher to determine if a block proposal is slashable, there is no lock around this request and can lead to race conditions. 

Here is the high level flow:
- Validator makes gRPC request IsSlashableBlock x3
- Slasher checks this request is not slashable by looking in the database
- Slasher updates the database with the slashing request

The problem here is that, if there is no lock then the condition at step 2 can be true when step 3 has not occurred yet for some other parallel request.

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
linux amd64
  </code>
</pre>

**What version of Prysm are you running? (Which release)**

<pre>
  <code>
alpha.18
  </code>
</pre>

**Anything else relevant (validator index / public key)?**

https://github.com/goerli/medalla/issues/37

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

 Bug fix

**What does this PR do? Why is it needed?**
Race condition makes slasher protection miss slashing conditions
this pr aim to add lock to avoid this issue

**Which issues(s) does this PR fix?**

Fixes #6868

### CODE REVIEW

File: slasher/rpc/server.go [hunk 1]
```diff
@@ -108,6 +113,8 @@ func (ss *Server) IsSlashableAttestation(ctx context.Context, req *ethpb.Indexed
 func (ss *Server) IsSlashableBlock(ctx context.Context, req *ethpb.SignedBeaconBlockHeader) (*slashpb.ProposerSlashingResponse, error) {
 	ctx, span := trace.StartSpan(ctx, "detection.IsSlashableBlock")
 	defer span.End()
+	ss.proposeLock.Lock()
+	defer ss.proposeLock.Unlock()
```
Comment (prestonvanloon) [inline slasher/rpc/server.go:117]: One thing to consider is that this is a global lock on this method. High QPS slashers will have a bit of lock contention.

Perhaps we could add some type of multi lock? I think we should consider it for a follow up PR though.

Comment (shayzluf) [reply slasher/rpc/server.go:117]: opening an issue for that

File: slasher/rpc/server.go [hunk 2]
```diff
@@ -108,6 +113,8 @@ func (ss *Server) IsSlashableAttestation(ctx context.Context, req *ethpb.Indexed
...
```
Comment (0xKiwi) [inline slasher/rpc/server.go:118]: Do the locks have to be this early? Wouldn't it be better to have them under the nil checks?

Comment (shayzluf) [reply slasher/rpc/server.go]: moved them down , thanks

**Pull request merged**